src/rmarkdown/preview.ts

```diff
@@ -0,0 +1,328 @@
+    // has a few idiosyncracies with view columns due to some limitations with
+    // vscode api. the view column will be set in order of priority:
+    //    1. the original document's view column when the preview button was pressed
+    //    2. the current webview's view column
+    //    3. the current active editor
+    // this is because we cannot tell the view column of a file if it is not visible
+    // (e.g., is an unopened tab)
+    public async showSource(): Promise<void> {
+        if (this.activePreview) {
+            await vscode.commands.executeCommand('vscode.open', this.activePreview.uri, {
+                preserveFocus: false,
+                preview: false,
+                viewColumn: this.activePreview.resourceViewColumn ?? this.activePreview.panel.viewColumn ?? vscode.ViewColumn.Active
+            });
+        }
+    }
+
+    public openExternalBrowser(): void {
+        if (this.activePreview) {
+            void vscode.env.openExternal(this.activePreview.outputUri);
+        }
+    }
+
+    private async retrieveSpawnData(fileUri: vscode.Uri, fileName: string, viewer: vscode.ViewColumn, currentViewColumn: vscode.ViewColumn, uri?: vscode.Uri) {
+        this.busyUriStore.add(fileUri);
+        await doWithProgress(async (token: vscode.CancellationToken) => {
+            await this.spawnProcess(fileUri, fileName, token, viewer, currentViewColumn);
+        },
+            vscode.ProgressLocation.Notification,
+            `Knitting ${fileName}...`,
+            true
+        ).catch((rejection: {
+            cp: cp.ChildProcessWithoutNullStreams,
+            wasCancelled?: boolean
+        }) => {
+            if (!rejection.wasCancelled) {
+                void vscode.window.showErrorMessage('There was an error in knitting the document. Please check the R Markdown output stream.');
+                this.rMarkdownOutput.show(true);
+            }
+            // this can occur when a successfuly knitted document is later altered (while still being previewed)
+            // and subsequently fails to knit
+            if (this.childStore.has(uri)) {
+                this.childStore.delete(this.childStore.get(uri));
+            } else {
+                rejection.cp.kill('SIGKILL');
+            }
+        });
+        this.busyUriStore.delete(fileUri);
+    }
+
+    private async spawnProcess(fileUri: vscode.Uri, fileName: string, token?: vscode.CancellationToken, viewer?: vscode.ViewColumn, currentViewColumn?: vscode.ViewColumn) {
+        return await new Promise<cp.ChildProcessWithoutNullStreams>((resolve, reject) => {
+            const lim = '---vsc---';
+            const re = new RegExp(`.*${lim}(.*)${lim}.*`, 'ms');
+            const cmd = (
+                `${this.rPath} --silent --slave --no-save --no-restore -e ` +
+                `"cat('${lim}', rmarkdown::render('${String(fileUri.path)}', output_dir = '${this.watcherDir}'), '${lim}', sep=''); while(TRUE) Sys.sleep(1)" `
```

Comment:
Using `~/.vscode-R` to store the output files might not be a good idea. If two Rmd files have the same name but in different folders, they will have the same output file.

I suggest using system temp dir (`os.tmpdir()`) instead. Since the R session will be killed each time before update, the R session `tempdir()` does not work with the next R session. Therefore, we might need to create a temp dir for vscode-R and each file has a unique output file (a random file name or a `fileUri` hash or something). Also, we might also better cleanup the temp files we created on some point (e.g. once the file is read?)

Reply:
Just notice that the output file will be removed.

Also, just remind that we use `~/.vscode-R` instead of `/tmp/vscode-R` to avoid user clash on the same directory.